pkg/api/vm/store.go

```diff
@@ -0,0 +1,62 @@
+package vm
+
+import (
+	"fmt"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/rancher/apiserver/pkg/types"
+	ctlcdiv1alpha1 "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1alpha1"
+	ctlkubevirtv1alpha3 "github.com/rancher/harvester/pkg/generated/controllers/kubevirt.io/v1alpha3"
+)
+
+type vmStore struct {
+	types.Store
+
+	vmCache     ctlkubevirtv1alpha3.VirtualMachineCache
+	dataVolumes ctlcdiv1alpha1.DataVolumeClient
+}
+
+func (s *vmStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	deleteDisk := request.Query.Get("deleteDisk")
+	if deleteDisk == "false" {
+		if err := s.updateDataVolumeOwnerRef(request.Namespace, request.Name); err != nil {
+			return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to update dataVolume's ownerReferences, %v", err))
+		}
+	}
+
+	return s.Store.Delete(request, request.Schema, id)
```

Comment:
This assumes that DVs are defined in VM spec. It does not consider the case when DV is created independently and the users want to remove them on VM deletion.